File: ios/TimberCruisingApp/Common/Units.swift
// Spec §8 (Common/Units.swift) + REQ-CAL-001 (per-project unit system).
// Imperial ↔ metric conversions. Pure functions, no imports beyond Foundation.

import Foundation

public enum Units {

    // MARK: Length — meters ↔ feet
    public static let metersPerFoot: Double = 0.3048
    public static func metersToFeet(_ m: Double) -> Double { m / metersPerFoot }
    public static func feetToMeters(_ ft: Double) -> Double { ft * metersPerFoot }

    // MARK: Length — cm ↔ inches
    public static let cmPerInch: Double = 2.54
    public static func cmToInches(_ cm: Double) -> Double { cm / cmPerInch }
    public static func inchesToCm(_ inches: Double) -> Double { inches * cmPerInch }

    // MARK: Area — m² ↔ ft²
    public static func squareMetersToSquareFeet(_ m2: Double) -> Double {
        m2 * (1.0 / metersPerFoot) * (1.0 / metersPerFoot)
    }
    public static func squareFeetToSquareMeters(_ ft2: Double) -> Double {
        ft2 * metersPerFoot * metersPerFoot
    }

    // MARK: Area — acres ↔ m²
    // 1 acre = 43560 ft² (exact) = 4046.8564224 m² (exact by definition of intl. foot)
    public static let squareMetersPerAcre: Double = 4046.8564224
    public static func acresToSquareMeters(_ ac: Double) -> Double { ac * squareMetersPerAcre }
    public static func squareMetersToAcres(_ m2: Double) -> Double { m2 / squareMetersPerAcre }

    // MARK: Volume — m³ ↔ ft³
    public static func cubicMetersToCubicFeet(_ m3: Double) -> Double {
        m3 * (1.0 / metersPerFoot) * (1.0 / metersPerFoot) * (1.0 / metersPerFoot)
    }
    public static func cubicFeetToCubicMeters(_ ft3: Double) -> Double {
        ft3 * metersPerFoot * metersPerFoot * metersPerFoot
    }

    // MARK: Per-acre basal area m²/ha ↔ ft²/ac (useful convenience)
    // 1 m²/ha = 4.35600 ft²/ac (derived from 1 ha = 2.47105 ac, 1 m² = 10.7639 ft²)
    public static func baPerHaToBaPerAcre(_ m2PerHa: Double) -> Double {
        m2PerHa * (squareMetersToSquareFeet(1.0) / 2.4710538147)
    }
    public static func baPerAcreToBaPerHa(_ ft2PerAc: Double) -> Double {
        ft2PerAc * (2.4710538147 / squareMetersToSquareFeet(1.0))
    }
}

// Float convenience overloads (spec uses Float for most measurement fields).
public extension Units {
    static func metersToFeet(_ m: Float) -> Float { Float(metersToFeet(Double(m))) }
    static func feetToMeters(_ ft: Float) -> Float { Float(feetToMeters(Double(ft))) }
    static func cmToInches(_ cm: Float) -> Float { Float(cmToInches(Double(cm))) }
    static func inchesToCm(_ inches: Float) -> Float { Float(inchesToCm(Double(inches))) }
    static func acresToSquareMeters(_ ac: Float) -> Float { Float(acresToSquareMeters(Double(ac))) }
    static func squareMetersToAcres(_ m2: Float) -> Float { Float(squareMetersToAcres(Double(m2))) }
}

File: android/app/src/main/java/com/hcjeong/forestix/common/Units.kt
// Imperial <-> metric conversions. Direct port of iOS Common/Units.swift.
// Pure functions, no platform dependencies.

package com.hcjeong.forestix.common

import kotlin.math.PI

/// Per-project unit system (REQ-CAL-001). Storage is always metric; this
/// only governs display + export rendering.
enum class UnitSystem { METRIC, IMPERIAL }

object Units {
    const val METERS_PER_FOOT = 0.3048
    fun metersToFeet(m: Double) = m / METERS_PER_FOOT
    fun feetToMeters(ft: Double) = ft * METERS_PER_FOOT

    /// Breast height, US convention: 4.5 ft written in metres.
    ///
    /// THE ONE PLACE THIS NUMBER LIVES ON ANDROID. Anything that draws,
    /// labels or reasons about breast height reads it from here, so the
    /// convention can be changed in a single edit. Kept equal to the iOS
    /// `Units.breastHeightM` — one rule, two platforms.
    ///
    /// It is NOT the 1.3 m that appears throughout the inventory engine
    /// (Näslund's asymptote in the H–D model, the `h > 1.3` volume guards,
    /// Laasasenaho's `(h − 1.3)` form). Those 1.3s belong to published
    /// equations and mean the international convention; retuning them to
    /// this value would move every fitted height and every volume in the
    /// corpus, which is a study-level decision, not a constant swap.
    const val BREAST_HEIGHT_M = 1.37

    const val CM_PER_INCH = 2.54
    fun cmToInches(cm: Double) = cm / CM_PER_INCH
    fun inchesToCm(inches: Double) = inches * CM_PER_INCH

    fun squareMetersToSquareFeet(m2: Double) = m2 * (1.0 / METERS_PER_FOOT) * (1.0 / METERS_PER_FOOT)
    fun squareFeetToSquareMeters(ft2: Double) = ft2 * METERS_PER_FOOT * METERS_PER_FOOT

    const val SQUARE_METERS_PER_ACRE = 4046.8564224
    fun acresToSquareMeters(ac: Double) = ac * SQUARE_METERS_PER_ACRE
    fun squareMetersToAcres(m2: Double) = m2 / SQUARE_METERS_PER_ACRE

    /// 1 hectare = 2.4710538147 acres (from the exact international-foot acre).
    /// Canonical single source for per-acre → per-hectare density conversion —
    /// AreaUnit multiplies a canonical per-acre figure by this to express it
    /// per hectare, and divides an area in acres by it to get hectares.
    const val ACRES_PER_HECTARE = 2.4710538147

    fun cubicMetersToCubicFeet(m3: Double) =
        m3 * (1.0 / METERS_PER_FOOT) * (1.0 / METERS_PER_FOOT) * (1.0 / METERS_PER_FOOT)

    fun circleAreaM2(radiusM: Double) = PI * radiusM * radiusM
}
